### PR TITLE
[SQL Server] Typed literals for date/time types

### LINKED ISSUE #1211: SQL Server datetime inline
`DateTime` parameter is inlined without `convert` for SQL Server. In general this is unsafe.

### Environment details
linq2db version: *2.0.0*
Database Server: *SQL Server 2017*
Operating system: *E.g. Windows 10*
Framework version: *.NET Core 2.0*

### LINKED ISSUE #3570: DateTime conversion missing when using external values
### Issue and steps to reproduce

I have a table joining with a range of dates which is from `SelectRangeIterator`. The LINQ query is like:

```c#
var startDate = new DateTime(2022, 5, 15);
var dates = Enumerable.Range(0, 7).Select(x => startDate.AddDays(x));

var result = from s in source    // the table
	join d in dates on 1 equals 1
	where d >= p.StartDate && d <= p.EndDate
	select new
	{
		p.WhatEver,
		Date = d
	};
```

And the translated SQL query is like:

```sql
SELECT
	[x].[WhatEver],
	[d].[item],
FROM
	[Source] [x]
		CROSS JOIN (VALUES
			('2022-05-15T00:00:00.0000000'), ('2022-05-16T00:00:00.0000000'),
			('2022-05-17T00:00:00.0000000'), ('2022-05-18T00:00:00.0000000'),
			('2022-05-19T00:00:00.0000000'), ('2022-05-20T00:00:00.0000000'),
			('2022-05-21T00:00:00.0000000')
		) [d]([item])
WHERE
	[d].[item] >= [x].[StartDate] AND
	[d].[item] <= [x].[EndDate])
```

It will throw a `Microsoft.Data.SqlClient.SqlException (0x80131904)`. I guess the problem is that the datetime2 conversion is missing for the `WHERE` clause, while implicit conversion fails.

That should be like:

```sql
WHERE
	Convert(DateTime2, [d].[item], 120) >= [x].[StartDate] AND
	Convert(DateTime2, [d].[item], 120) <= [x].[EndDate])
```

### Exception

```
Microsoft.Data.SqlClient.SqlException (0x80131904): Conversion failed when converting date and/or time from character string.
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlDataReader.TryHasMoreRows(Boolean& moreRows)
   at Microsoft.Data.SqlClient.SqlDataReader.TryReadInternal(Boolean setTimeout, Boolean& more)
   at Microsoft.Data.SqlClient.SqlDataReader.Read()
   at LinqToDB.Linq.QueryRunner.ExecuteElement[T](Query query, IDataContext dataContext, Mapper`1 mapper, Expression expression, Object[] ps, Object[] preambles)```
```

### Environment details

linq2db version: 3.7.0
Database Server: SQL Server 2014
Operating system: Windows 11 Pro
.NET Framework: .NET 6

### PR DESCRIPTION
**Pull request by MaceWindu**
Fix #3570, Fix #1211 and similar issues where sql server cannot type string as date/time type

### CODE REVIEW

**Comment by jods4** [comment]
May be worth investigating what makes #3570 fail.
This identical fiddle works:
https://dbfiddle.uk/?rdbms=sqlserver_2017l&fiddle=be7227f9cd3dc05ff2a91a7d4de1e4d4

**Comment by wanaxe** [comment]
> May be worth investigating what makes #3570 fail. This identical fiddle works: https://dbfiddle.uk/?rdbms=sqlserver_2017l&fiddle=be7227f9cd3dc05ff2a91a7d4de1e4d4

Thank you so much for looking into the issue. The #3570 fails because the date column in my table is datetime (not datetime2).

If changing the datetime2 to datetime in the fiddle, you'll see the error, as implict coversion (from 'yyyy-MM-ddTHH:mm:ss.fffffff' to datetime) fails

**Review by viceroypenguin — APPROVED** [review]

**Comment by jods4** [comment]
A simpler change could be to use literal format `20220501 00:00:00.000`, it's supposed to work with every date type I think, fiddle:
https://dbfiddle.uk/?rdbms=sqlserver_2017l&fiddle=2e2b29ca5f7a7d19e1c5a7a3dfcf50ae

It doesn't work with precision past millisecond, not sure if that's an issue.

**Review by sdanyliv — APPROVED** [review]

File: Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs [hunk 1]
```diff
@@ -14,34 +14,83 @@ namespace LinqToDB.DataProvider.SqlServer
...
+			"'{0:yyyy-MM-ddTHH:mm:ss.fffffffzzz}'",
+		};
+		private static readonly string[] DATETIMEOFFSET_TYPED_FORMATS = new[]
+		{
+			"CAST('{0:yyyy-MM-ddTHH:mm:sszzz}' AS DATETIMEOFFSET(0))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.fzzz}' AS DATETIMEOFFSET(1))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.ffzzz}' AS DATETIMEOFFSET(2))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.fffzzz}' AS DATETIMEOFFSET(3))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.ffffzzz}' AS DATETIMEOFFSET(4))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.fffffzzz}'} AS DATETIMEOFFSET(5))",
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs:90]:
```suggestion
			"CAST('{0:yyyy-MM-ddTHH:mm:ss.fffffzzz}' AS DATETIMEOFFSET(5))",
```

File: Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs [hunk 2]
```diff
@@ -14,34 +14,83 @@ namespace LinqToDB.DataProvider.SqlServer
...
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.ffffffzzz}' AS DATETIMEOFFSET(6))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.fffffffzzz}' AS DATETIMEOFFSET)"
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs:92]:
```suggestion
			"CAST('{0:yyyy-MM-ddTHH:mm:ss.fffffffzzz}' AS DATETIMEOFFSET)",
```

File: Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs [hunk 3]
```diff
@@ -14,34 +14,83 @@ namespace LinqToDB.DataProvider.SqlServer
...
+		private static readonly string[] DATETIME2_TYPED_FORMATS       = new[]
+		{
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss}' AS DATETIME2(0))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.f}' AS DATETIME2(1))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.ff}' AS DATETIME2(2))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.fff}' AS DATETIME2(3))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.ffff}' AS DATETIME2(4))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.fffff}' AS DATETIME2(5))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.ffffff}' AS DATETIME2(6))",
+			"CAST('{0:yyyy-MM-ddTHH:mm:ss.fffffff}' AS DATETIME2)"
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs:68]:
```suggestion
			"CAST('{0:yyyy-MM-ddTHH:mm:ss.fffffff}' AS DATETIME2)",
```

File: Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs [hunk 4]
```diff
@@ -176,82 +218,194 @@ static void ConvertCharToSql(StringBuilder stringBuilder, SqlDataType sqlDataTyp
...
+					break;
+				case (DataType.Text, _, _) or (DataType.Char, _, _) or (DataType.VarChar, _, _):
+					stringBuilder.AppendFormat(CultureInfo.InvariantCulture, DATE_FORMAT, value);
+					break;
+
+				case (_, false, _):
+					ConvertDateTimeToSql(stringBuilder, sqlDataType, value.ToDateTime(default), v2008plus, supportsFromParts);
+					break;
+
+				default:
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs:333]: should use `case (_, _, false):` and `case (_, _, true):` here. Let the compiler tell you if you've handled all of the cases

File: Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs [hunk 5]
```diff
@@ -176,82 +218,194 @@ static void ConvertCharToSql(StringBuilder stringBuilder, SqlDataType sqlDataTyp
...
+					stringBuilder.Append('N');
+					stringBuilder.AppendFormat(CultureInfo.InvariantCulture, DATETIMEOFFSET_FORMATS[precision], value);
+					break;
+				}
+
+				case (_, false, _) or (DataType.Date, _, _) or (DataType.DateTime, _, _) or (DataType.DateTime2, _, _) or (DataType.SmallDateTime, _, _):
+					ConvertDateTimeToSql(stringBuilder, sqlDataType, value.DateTime, v2008plus, supportsFromParts);
+					return;
+
+				default:
```
Comment (viceroypenguin) [inline Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs:374]: Should split out to `case (_, _, false):` and `case (_, _, true):` here. Let the compiler tell you if you've handled all of the cases

File: Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs (hunk 1 above)
Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs:90]: looks like don't have enough coverage here... will add more tests

File: Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs (hunk 4 above)
Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/SqlServer/SqlServerMappingSchema.cs:333]: are you sure? `(_, _, false) + (_, _, true)` already cover all cases

**Review by sdanyliv — DISMISSED** [review]

**Pull request merged**